### PR TITLE
Speed up local image builds by defining Skaffold dependencies

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity
/kind enhancement

**What this PR does / why we need it**:
#6735 already started with it, however it actually introduced more problems than it solved since it didn't specify the full list of dependencies. Consequently, the images were only re-built when there were changes in the provided packages. If there were changes in other packages (actually used but not provided in `skaffold.yaml`) then they were not re-built.

With this PR, we define all transitive `gardener/gardener` package dependencies for each binary which (a) fixes the mentioned problem and (b) speeds up image builds in general.

The information was retrieved via

```shell
go list -f '{{ .Deps }}' ./cmd/gardener-apiserver
```

etc.

**Which issue(s) this PR fixes**:
Part of #6016

**Special notes for your reviewer**:
As you can see, there are quite some weird dependencies that shouldn't exist ideally (e.g., `gardener-apiserver` should not depend on `pkg/resourcesmanager/*`, etc.). Improving this is out of scope of this PR, however I'll add a respective task to #6016.

**Release note**:

```other operator
NONE
```

### CODE REVIEW

**Comment by plkokanov** [comment]
This means that any new package that we add will also have to be added to the appropriate dependency list, right? Is there anyway we can add a check whether that was done, or document it?

**Comment by rfranzke** [comment]
Fair point @plkokanov, I have reworked the PR and added a script that checks for the correctly defined dependencies in the `skaffold.yaml` as part of `make check`. As a result, it's much more verbose now, but definitely much easier and less error-prone to maintain. Thanks for the suggestion, PTAL.

File: hack/check-skaffold-deps.sh [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  path_current_skaffold_dependencies="${out_dir}/current-skaffold-deps-$binary_name.txt"
+  path_actual_dependencies="${out_dir}/actual-deps-$binary_name.txt"
+
+  echo "$skaffold_yaml" |\
+    yq eval "select(.metadata.name == \"$skaffold_config_name\") | .build.artifacts[] | select(.ko.main == \"./cmd/$binary_name\") | .ko.dependencies.paths[]?" - |\
+    sort |\
+    uniq > "$path_current_skaffold_dependencies"
+
+  go list -f '{{ .Deps }}' "./cmd/$binary_name" |\
+    tr " " "\n" |\
```
Comment (plkokanov) [inline hack/check-skaffold-deps.sh]: You can skip the `tr` call by using 
```
  go list -f '{{join .Deps "\n"}}' "./cmd/$binary_name" |\
```

**Review by plkokanov — CHANGES_REQUESTED** [review]
Thanks for adding the check!
Just one very small nit/suggestion, otherwise lgtm.

**Pull request merged**